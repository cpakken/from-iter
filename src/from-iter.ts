import { IterBase, PrevResultStore } from './IterBase'

export class IterList<T> extends IterBase<T, number> {
  protected _iterate(callback: (result: T, key: number, index: number) => void) {
    let key = 0
    let index = 0

    //Store previous mapReduce results per reducer
    const prevResultStore: PrevResultStore<T, number> = new WeakMap()

    for (const item of this._iterator) {
      const state = this._pipeItem(callback, prevResultStore, item, key, index)
      if (state > 0) index++
      if (state < 0) break
      key++
    }
  }
}

export type IterObjectKey<ITER> = ITER extends IterObject<any, infer KEY> ? KEY : never

export class IterObject<T, KEY extends string = string> extends IterBase<T, KEY> {
  protected _iterate(callback: (result: T, key: KEY, index: number) => void) {
    let index = 0

    //Store previous mapReduce results per reducer
    const prevResultStore: PrevResultStore<T, string> = new WeakMap()
    const object = this._iterator as any

    for (const key in object) {
      const item = object[key]
      const state = this._pipeItem(callback, prevResultStore, item, key as KEY, index)
      if (state > 0) index++
      if (state < 0) break
    }
  }
}

export function fromIter<T>(iterator: Iterable<T>): IterList<T>
export function fromIter<T extends { [key: string]: any }>(
  iterator: T
): IterObject<T[keyof T], keyof T & string>
export function fromIter(iterator: any) {
  return iterator[Symbol.iterator] ? new IterList(iterator) : new IterObject(iterator)
}
