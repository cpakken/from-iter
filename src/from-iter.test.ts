import { Mock } from 'vitest'
import { fromIter } from './from-iter'

const getMockResults = (mock: Mock) => mock.mock.results.map((r) => r.value)

describe('IterList', () => {
  const numbers = Array.from({ length: 10 }).map((_, i) => i)

  test('map', () => {
    const list = fromIter(numbers)

    const mapped = list.map((value, key) => ({ value, key }))

    const array = mapped.toArray()

    expect(array).toMatchSnapshot()
  })
})

describe('IterObject', () => {
  const object = { foo: 'bar', baz: 'qux' }

  test('map', () => {
    const list = fromIter(object)

    const result = list.map((value, key) => key.concat(value))

    expect(result.toArray()).toEqual(['foobar', 'bazqux'])

    expect(result.toObject()).toEqual({ foo: 'foobar', baz: 'bazqux' })
  })
})

describe('kitchen sink', () => {
  const mapKeys = vi.fn((_value, key: number) => key)

  const addPrev = vi.fn((prev: number, current: number) => prev + current)

  const filterEven = vi.fn((value) => value % 2 === 0)

  const stopAt = vi.fn((value: number) => value > 100)

  const mapResults = vi.fn((value: number, key: number, index: number) => ({ value, key, index }))

  const a = fromIter(Array.from({ length: 20 }))
    .map(mapKeys)
    .mapReduce(addPrev, 0)
    .filter(filterEven)
    .until(stopAt)
    .map(mapResults)

  beforeEach(() => {
    vi.clearAllMocks()
  })

  test('toArray', () => {
    const list = a.toArray()
    expect(list).toMatchSnapshot()

    expect(getMockResults(mapKeys)).toEqual(Array.from({ length: 16 }).map((_, i) => i))
    expect(getMockResults(addPrev)).toMatchSnapshot()
    expect(getMockResults(filterEven)).toMatchSnapshot()
  })

  test('reduce', () => {
    const result = a.reduce((acc, { value, key }) => ((acc[key] = value), acc), {} as any)

    expect(result).toMatchInlineSnapshot(`
      {
        "0": 0,
        "11": 66,
        "12": 78,
        "3": 6,
        "4": 10,
        "7": 28,
        "8": 36,
      }
    `)
  })
})
