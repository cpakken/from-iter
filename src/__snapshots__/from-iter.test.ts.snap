// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`IterList > map 1`] = `
[
  {
    "key": 0,
    "value": 0,
  },
  {
    "key": 1,
    "value": 1,
  },
  {
    "key": 2,
    "value": 2,
  },
  {
    "key": 3,
    "value": 3,
  },
  {
    "key": 4,
    "value": 4,
  },
  {
    "key": 5,
    "value": 5,
  },
  {
    "key": 6,
    "value": 6,
  },
  {
    "key": 7,
    "value": 7,
  },
  {
    "key": 8,
    "value": 8,
  },
  {
    "key": 9,
    "value": 9,
  },
]
`;

exports[`kitchen sink > toArray 1`] = `
[
  {
    "index": 0,
    "key": 0,
    "value": 0,
  },
  {
    "index": 1,
    "key": 3,
    "value": 6,
  },
  {
    "index": 2,
    "key": 4,
    "value": 10,
  },
  {
    "index": 3,
    "key": 7,
    "value": 28,
  },
  {
    "index": 4,
    "key": 8,
    "value": 36,
  },
  {
    "index": 5,
    "key": 11,
    "value": 66,
  },
  {
    "index": 6,
    "key": 12,
    "value": 78,
  },
]
`;

exports[`kitchen sink > toArray 2`] = `
[
  0,
  1,
  3,
  6,
  10,
  15,
  21,
  28,
  36,
  45,
  55,
  66,
  78,
  91,
  105,
  120,
]
`;

exports[`kitchen sink > toArray 3`] = `
[
  true,
  false,
  false,
  true,
  true,
  false,
  false,
  true,
  true,
  false,
  false,
  true,
  true,
  false,
  false,
  true,
]
`;
